### ISSUE TITLE
[Discussion] What if we get rid of static modules?

### ISSUE THREAD

**Issue by ericmorand** [nestjs/nest#6375]
I have been using NestJS for a few weeks now and let me say that I truly love it. Especially the DI features.

But the more I use it, the more I tend to diverge from the documentation and find some (subjectively) more consistent approach to write NestJS applications. I started that sentence with "But" but it is actually the thing I love the most with NestJS: it's so well thought and written that it works whatever the way you write things, as long as you match the specification (aka the typing mostly).

One of the thing that I now do is getting rid of static modules and using dynamic modules all the time. I found that it made my code more consistent, easier to refactor and thus more mantainable. I also don't use static class methods to create dynamic modules but static _ES6 module_ constants instead.

For example, I now always write things like that:

```ts
import {createModule as createFooModule} from "./foo.module.ts";
import {MyProvider} from "./my/provider.ts";

export const MyModule: DynamicModule = {
    imports: [createFooModule()],
    providers: [MyProvider]
};
```
What I love with this approach is that it is consistent: it always relies on a symbol, instead of sometimes relying on a static module (by importing it class) and sometimes on dynamic module (by calling a static method on a class).

Another thing that I find disturbing with the Dynamic Module approach of the documentation is that it exposes a class, but expect the consumer to _know_ that this class should not be used as a module.

```ts
@Module({})
class FooModule {
    static register(): DynamicModule {
        // ...
    }
}
```

Using `FooModule` as the target of the `imports` property of another module would lead to a crash (because some dependencies are missing). The consumer has to _know_ that it needs to call the static `register` method to get a proper module. 

I believe this goes against the IoC principle where the consumer should not have to know the details of the construction of the dependency - and it's also not consistent with some static modules that can be imported using their class.

Basically, I'm not using static modules anymore. Because they don't provide anything valuable compared to dynamic modules (I even suspect that they are transformed into dynamic modules at some point by the application boostraping) and create inconsistencies.

What do you think? Is there a flaw that I don't know about with my approach? Could it be generalized as a "good practice" approach?

**Comment by kamilmysliwiec** [comment]
> Another thing that I find disturbing with the Dynamic Module approach of the documentation is that it exposes a class, but expect the consumer to know that this class should not be used as a module.

This is correct and it's something I'd definitely love to improve at some point in the future. Basically, a simple mechanism that would let define the target module that it shouldn't be used in a static way. Currently, to accomplish something like this, you'd have to define a provider, as follows:

```typescript
providers: [{
  provide: 'RANDOM_TOKEN',
  useFactory: () => throw new Error(`You shouldn't use the "X.register()" method when importing X module.`),
}]
```

And then overriding this provider within the `register()` method (so it doesn't throw an exception). This requires way too much boilerplate though so obviously we should provide a better, simpler solution to that.

> Because they don't provide anything valuable compared to dynamic modules (I even suspect that they are transformed into dynamic modules at some point by the application boostraping) and create inconsistencies.

This does not happen 😄 

> What do you think? Is there a flaw that I don't know about with my approach? Could it be generalized as a "good practice" approach?

Your approach is completely fine. A dynamic module definition is just a plain object (as you've noticed as well), so you can use whatever you want to construct them.

> What if we get rid of static modules?

There are no plans to drop static modules and I don't see any benefits in this tbh (+ that would introduce a huge breaking change, breaking pretty much all the existing packages and most of the applications).

**Comment by ericmorand** [comment]
> This requires way too much boilerplate though so obviously we should provide a better, simpler solution to that.

Maybe just updating the documentation to propose an alternative approach to static method? Maybe I can provide a PR...

Anyway, thanks a lot for the clarification and your time.

**Comment by kamilmysliwiec** [comment]
Let's track this here https://github.com/nestjs/docs.nestjs.com/issues/1741

**Comment by Ginden** [comment]
> Basically, a simple mechanism that would let define the target module that it shouldn't be used in a static way

It can be achieved by convention of adding static property `alwaysDynamic`.

Eg.

```
@Module({})
class SomeDynamicModule {
   static isDynamic = true;
   static register(): DynamicModule: 
}
```

**Comment by viglucci** [comment]
@ericmorand 

When I apply the same approach as your example, it doesn't work for me.

```ts
export const SlackBotModule: DynamicModule = {
  imports: [],
  providers: [],
};
```

The above gives the following type of error:

```
const SlackBotModule: DynamicModule
Property 'module' is missing in type '{ imports: never[]; providers: never[]; }' but required in type 'DynamicModule'.ts(2741)
```

Did you overcome this somehow, or have the types just changed since you used this pattern?
